explorer/interpreter/interpreter.cpp

```diff
@@ -860,6 +865,42 @@ auto Interpreter::Convert(Nonnull<const Value*> value,
       }
       return Convert(value, destination_type, source_loc);
     }
+    case Value::Kind::PointerValue: {
+      if (destination_type->kind() != Value::Kind::PointerType ||
+          cast<PointerType>(destination_type)->type().kind() !=
+              Value::Kind::NominalClassType) {
+        // No conversion needed.
+        return value;
+      }
+
+      // Get pointed value.
+      const auto* src_ptr = cast<PointerValue>(value);
+      CARBON_ASSIGN_OR_RETURN(const auto* pointee,
+                              heap_.Read(src_ptr->address(), source_loc))
+      if (pointee->kind() == Value::Kind::NominalClassType) {
+        // TODO: When does that happen? Just return as-is.
+        return value;
+      }
+
+      const auto* dest_ptr = cast<PointerType>(destination_type);
+      std::optional<Nonnull<const NominalClassValue*>> class_subobj =
+          cast<NominalClassValue>(pointee);
+      auto new_addr = src_ptr->address();
+      while (class_subobj) {
+        if (TypeEqual(&class_subobj.value()->type(), &dest_ptr->type(),
```

Comment:
I think we should prefer to use the pointer-like syntax for optionals where practical, because it's more concise, and makes it easier to distinguish the application logic from the mechanics of wrapping and unwrapping. It's also probably more consistent with how the corresponding Carbon code will look, and experimenting with writing Carbon-like code is the whole point of using `optional<Nonnull<T*>>` instead of `T*`.

```suggestion
        if (TypeEqual(&(*class_subobj)->type(), &dest_ptr->type(),
```